core/state/state_object.go

```diff
@@ -390,41 +390,57 @@ func (s *StateObject) updateTrie(db Database) Trie {
 			s.db.MetricsMux.Unlock()
 		}(time.Now())
 	}
-	// The snapshot storage map for the object
-	var storage map[string][]byte
 	// Insert all the pending updates into the trie
 	tr := s.getTrie(db)
 
 	usedStorage := make([][]byte, 0, len(s.pendingStorage))
+	dirtyStorage := make(map[common.Hash][]byte)
 	for key, value := range s.pendingStorage {
 		// Skip noop changes, persist actual changes
 		if value == s.originStorage[key] {
 			continue
 		}
 		s.originStorage[key] = value
 		var v []byte
-		if (value == common.Hash{}) {
-			s.setError(tr.TryDelete(key[:]))
-		} else {
+		if (value != common.Hash{}) {
 			// Encoding []byte cannot fail, ok to ignore the error.
 			v, _ = rlp.EncodeToBytes(common.TrimLeftZeroes(value[:]))
-			s.setError(tr.TryUpdate(key[:], v))
 		}
+		dirtyStorage[key] = v
+	}
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for key, value := range dirtyStorage {
+			if len(value) == 0 {
+				s.setError(tr.TryDelete(key[:]))
+			} else {
+				s.setError(tr.TryUpdate(key[:], value))
+			}
+			usedStorage = append(usedStorage, common.CopyBytes(key[:]))
+		}
+	}()
+	if s.db.snap != nil {
 		// If state snapshotting is active, cache the data til commit
-		if s.db.snap != nil {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			// The snapshot storage map for the object
 			s.db.snapMux.Lock()
+			storage := s.db.snapStorage[s.address]
 			if storage == nil {
-				// Retrieve the old storage map, if available, create a new one otherwise
-				if storage = s.db.snapStorage[s.address]; storage == nil {
-					storage = make(map[string][]byte)
-					s.db.snapStorage[s.address] = storage
-				}
+				storage = make(map[string][]byte)
```

Comment:
storage = make(map[string][]byte, len(dirtyStorage))
since the capacity is known.

Reply:
updated